fastlane/lanes/test.rb

```diff
@@ -43,6 +43,49 @@
       UI.test_failure!(message)
     end
   end
+
+  #####################################################################################
+  # build_and_run_jetpack_instrumented_test
+  # -----------------------------------------------------------------------------------
+  # Run instrumented tests in Google Firebase Test Lab
+  # -----------------------------------------------------------------------------------
+  # Usage:
+  # bundle exec fastlane build_and_run_jetpack_instrumented_test
+  #
+  #####################################################################################
+  desc "Build the application and instrumented tests, then run the tests in Firebase Test Lab"
+  lane :build_and_run_jetpack_instrumented_test do | options |
```

Comment:
Suggestion (💡): This lane is an almost duplicate of the `build_and_run_wordpress_instrumented_test` above. The only diffs I am seeing are the below:
- lane:
    - `assembleWordPressVanillaDebug` and `assembleJetpackVanillaDebug` -> `WordPress` and `Jetpack`)
    - `assembleWordPressVanillaDebugAndroidTest` and `assembleJetpackVanillaDebugAndroidTest`  -> `WordPress` and `Jetpack`
- test_apk_path:
    - `wordpressVanilla` and `jetpackVanilla` -> `wordpress` and `jetpack`
    - `org.wordpress.android-wordpress-vanilla-debug-androidTest.apk` and `org.wordpress.android-jetpack-vanilla-debug-androidTest.apk` -> `wordpress` and `jetpack`
- apk_path:
    - `wordpressVanilla` and `jetpackVanilla` -> `wordpress` and `jetpack`
    - `org.wordpress.android-wordpress-vanilla-debug.apk` and `org.wordpress.android-jetpack-vanilla-debug.apk` -> `wordpress` and `jetpack`
- test_targets:
    - `notPackage org.wordpress.android.ui.screenshots` and `package org.wordpress.android.e2e` -> 👍
- annotation_ctx:
    - `firebase-test-wordpress-vanilla-debug` and `firebase-test-jetpack-vanilla-debug` -> `wordpress` and `jetpack`

From the above you extract the following pattern for de-duplication needs:
- `app` key for the `WordPress` and `Jetpack` values.
- Another, or the same `app` key, for the `wordPress` and `jetpack` values.
- `test_targets` key for its corresponding values.

I think it is important that we remove this duplication to avoid future problems when one of the lanes is updated without the other one being updated too, and this way, we'll also avoid maintaining more than we should.

As such, I suggest creating one parent lane (`build_and_run_instrumented_test`) which would hold parameters like `app` and `test_targets` and accordingly pass them to the children lanes (`build_and_run_wordpress_instrumented_test` and `build_and_run_jetpack_instrumented_test`).

PS: On the `app` key, I think we already have something like that in place. Maybe @AliSoftware can add some ideas on how to make my above de-duplication suggest more robust.

Reply:
+1 on [DRYing](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) the code.

As for tips, you could use the same technique as what we use [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/localization.rb#L137-L139). Basically:
 - [We have a Dictionary here which is defining a set of constants with their different values for WordPress vs Jetpack](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/Fastfile#L3-L24)
 - On our lanes that are in the same case as yours, i.e. we do mostly the same for WordPress vs Jetpack with only a small set of variations, we use our [`app = get_app_name_option!(options)` helper method](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/release.rb#L314-L321) to read the `app:` parameter that is passed to the lane (and normalize it and error nicely if it's missing), and then we use `APP_SPECIFIC_VALUES[app][:some_key]` to read the app-specific value for a given `some_key` constant defined from this dictionary

This allows you to avoid having to pass many parameters to the lane (all having to be consistent with one another) to invoke things for WordPress vs Jetpack, or having to create a parent lane accepting many parameters then add two children lanes invoking it with the right values (like Petros suggested and like we used to do a while ago before we started to use this `APP_SPECIFIC_VALUES` technique)… and instead you [just have to pass `app:wordpress` or `app:jetpack` when invoking those kind of lanes](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/localization.rb#L132-L134) and the implementation of your lane will pick the constants for the right app from the `APP_SPECIFIC_VALUES` constants, ensuring consistency in the different values being used for a given `app:`.

Reply:
Simplified code in these commits: https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/5b9a9fc827740a1cf89fdfeae62ac9b0529b5d5b and https://github.com/wordpress-mobile/WordPress-Android/pull/17950/commits/ea09f0407c8ddd1d38b4e5f91d8023a71b7b01ce (I missed one change in the first commit 😅)

I decided to use the `APP_SPECIFIC_VALUES` way that @AliSoftware suggested so there is only 1 lane that should now work for both apps.

Reply:
Awesome update @jostnes , Mr. @AliSoftware you are the best! 🌟 🌟 🌟